File: tks/ApplicationCore/ApplicationServices/src/test/kotlin/services/ResourcesServiceTest.java
package services;

import exceptions.IncompatibleResourceFormatException;
import exceptions.ResourceBlockedByRentException;
import exceptions.ResourceNotFoundException;
import exceptions.UnknownResourceException;
import drivenports.RentQueryPort;
import helpers.AccessionNumberHelper;
import lombok.SneakyThrows;
import model.values.Email;
import ports.secondary.ResourcePersistencePort;
import ports.secondary.ResourceSearchPort;
import model.*;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Date;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
public class ResourcesServiceTest {

    ResourcesService resourcesService;

    Resource sampleBook;
    Resource sampleMagazine;
    Resource invalidResource;

    @Mock
    ResourcePersistencePort resourcePersistencePort;
    @Mock
    ResourceSearchPort resourceSearchPort;
    @Mock
    RentQueryPort rentQueryPort;

    @BeforeEach
    public void init() {
        resourcesService = new ResourcesService(resourcePersistencePort, resourceSearchPort, rentQueryPort);
        sampleBook = new Book(null, "Diuna", "Frank Herbert");
        sampleMagazine = new Magazine(null, "Świerszczyk", "Nowa era");
        invalidResource = new InvalidResource(null, "invalid");
    }

    @Test
    public void GivenResourceOfValidType_Create_ShouldSuccess(){
        resourcesService.create(sampleBook);
        verify(resourcePersistencePort).add(eq(sampleBook));
        assertNotNull(sampleBook.getId());
    }

    @Test
    public void GivenResourceOfInvalidType_Create_ShouldFail(){
        assertThrows(UnknownResourceException.class, () -> resourcesService.create(invalidResource));
        verify(resourcePersistencePort, never()).add(any());
    }

    @Test
    public void GivenResourceWithNotNullId_Create_ShouldFail(){
        sampleBook.setId(AccessionNumberHelper.generate());
        assertThrows(UnknownResourceException.class, () -> resourcesService.create(sampleBook));
        verify(resourcePersistencePort, never()).add(any());
    }

    @Test
    public void GivenValidResourceWithNotNullId_Update_ShouldSuccess(){
        sampleMagazine.setId(AccessionNumberHelper.generate());
        when(resourceSearchPort.findById(eq(sampleMagazine.getId()))).thenReturn(sampleMagazine);
        resourcesService.update(sampleMagazine);
        verify(resourcePersistencePort).save(sampleMagazine);
    }

    @Test
    public void GivenNonExistingResource_Update_ShouldFail(){
        sampleMagazine.setId(AccessionNumberHelper.generate());
        when(resourceSearchPort.findById(eq(sampleMagazine.getId()))).thenReturn(null);
        assertThrows(ResourceNotFoundException.class, () -> resourcesService.update(sampleMagazine));
        verify(resourcePersistencePort, never()).save(any());
    }

    @Test
    public void GivenOtherResourceType_Update_ShouldFail(){
        sampleMagazine.setId(AccessionNumberHelper.generate());
        sampleBook.setId(sampleMagazine.getId());
        when(resourceSearchPort.findById(eq(sampleMagazine.getId()))).thenReturn(sampleMagazine);
        assertThrows(IncompatibleResourceFormatException.class, () -> resourcesService.update(sampleBook));
        verify(resourcePersistencePort, never()).save(any());
    }

    @Test
    public void GivenValidResourceId_Remove_ShouldSuccess(){
        var accessionNumber = AccessionNumberHelper.generate();
        sampleMagazine.setId(accessionNumber);
        when(resourceSearchPort.findById(eq(accessionNumber))).thenReturn(sampleMagazine);
        when(rentQueryPort.findActiveByResourceId(eq(accessionNumber))).thenReturn(null);
        resourcesService.remove(accessionNumber);
        verify(resourcePersistencePort).remove(sampleMagazine);
    }

    @Test
    public void GivenInvalidResourceId_Remove_ShouldFail(){
        var accessionNumber = AccessionNumberHelper.generate();
        when(resourceSearchPort.findById(eq(accessionNumber))).thenReturn(null);
        assertThrows(ResourceNotFoundException.class, () -> resourcesService.remove(accessionNumber));
        verify(resourcePersistencePort, never()).remove(any());
    }

    @SneakyThrows
    @Test
    public void GivenRentResourceId_Remove_ShouldFail(){
        var accessionNumber = AccessionNumberHelper.generate();
        sampleMagazine.setId(accessionNumber);
        when(resourceSearchPort.findById(eq(accessionNumber))).thenReturn(sampleMagazine);
        when(rentQueryPort.findActiveByResourceId(eq(accessionNumber))).thenReturn(new Rent(UUID.randomUUID(), new Date(), null, new Email("mszewc@edu.pl"), accessionNumber));
        assertThrows(ResourceBlockedByRentException.class, () -> resourcesService.remove(accessionNumber));
        verify(resourcePersistencePort, never()).remove(any());
    }

    @Test
    public void GivenValidResourceId_GetDetails_ShouldSuccess(){
        var accessionNumber = AccessionNumberHelper.generate();
        sampleMagazine.setId(accessionNumber);
        when(resourceSearchPort.findById(eq(sampleMagazine.getId()))).thenReturn(sampleMagazine);
        var result = resourcesService.getDetails(sampleMagazine.getId());
        verify(resourceSearchPort).findById(eq(sampleMagazine.getId()));
        assertEquals(sampleMagazine, result);
    }

    @Test
    public void GivenInvalidResourceId_GetDetails_ShouldFail(){
        var accessionNumber = AccessionNumberHelper.generate();
        when(resourceSearchPort.findById(eq(accessionNumber))).thenReturn(null);
        assertThrows(ResourceNotFoundException.class, () -> resourcesService.getDetails(accessionNumber));
    }
}

File: tks/core/services/src/test/kotlin/application/services/ResourcesServiceTest.kt
package application.services

import application.helpers.AccessionNumberHelper
import domain.exceptions.IncompatibleResourceFormatException
import domain.exceptions.ResourceBlockedByRentException
import domain.exceptions.ResourceNotFoundException
import domain.exceptions.UnknownResourceException
import domain.model.context.library.Book
import domain.model.context.library.Magazine
import domain.model.context.rents.Rent
import domain.model.context.library.Resource
import domain.model.values.AccessionNumber
import domain.model.values.Email
import io.mockk.every
import io.mockk.impl.annotations.RelaxedMockK
import io.mockk.junit5.MockKExtension
import io.mockk.verify
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import ports.secondary.ResourcePersistencePort
import ports.secondary.ResourceSearchPort
import java.util.*

@ExtendWith(MockKExtension::class)
class ResourcesServiceTest {

    data class InvalidResource(
        override var accessionNumber: AccessionNumber?,
        override var title: String,
        override var locked: Boolean
    ) : Resource

    private lateinit var resourcesService: ResourcesService
    private lateinit var sampleBook: Resource
    private lateinit var sampleMagazine: Resource
    private lateinit var invalidResource: Resource
    private val sampleAccessionNumber: AccessionNumber = AccessionNumberHelper.generate()

    @RelaxedMockK
    lateinit var resourcePersistencePort: ResourcePersistencePort

    @RelaxedMockK
    lateinit var resourceSearchPort: ResourceSearchPort

    @BeforeEach
    fun init() {
        resourcesService = ResourcesService(resourcePersistencePort, resourceSearchPort)
        sampleBook = Book(null, "Diuna", false, "Frank Herbert")
        sampleMagazine = Magazine(null, "Świerszczyk",false,  "Nowa era")
        invalidResource = InvalidResource(null, "invalid", false)
    }

    @Test
    fun `Given resource of valid type then create should success`() {
        resourcesService.create(sampleBook)
        verify(exactly = 1) { (resourcePersistencePort).save(sampleBook) }
        Assertions.assertNotNull(sampleBook.accessionNumber)
    }

    @Test
    fun `Given resource of invalid type then create should fail`() {
        Assertions.assertThrows(UnknownResourceException::class.java) {
            resourcesService.create(
                invalidResource
            )
        }
        verify(exactly = 0) { resourcePersistencePort.save(any()) }
    }

    @Test
    fun `Given resource with not null id then create should fail`() {
        sampleBook.accessionNumber = sampleAccessionNumber
        Assertions.assertThrows(UnknownResourceException::class.java) {
            resourcesService.create(sampleBook)
        }
        verify(exactly = 0) { resourcePersistencePort.save(any()) }
    }

    @Test
    fun `Given valid resource with not null id then update should success`() {
        sampleMagazine.accessionNumber = sampleAccessionNumber
        every { (resourceSearchPort.findByAccessionNumber(sampleAccessionNumber)) }.returns(sampleMagazine)
        resourcesService.update(sampleMagazine)
        verify(exactly = 1) { (resourcePersistencePort).save(sampleMagazine) }
    }

    @Test
    fun `Given non existing resource then update should fail`() {
        sampleMagazine.accessionNumber = sampleAccessionNumber
        every { resourceSearchPort.findByAccessionNumber(sampleMagazine.accessionNumber!!) } returns null
        Assertions.assertThrows(ResourceNotFoundException::class.java) {
            resourcesService.update(sampleMagazine)
        }
        verify(exactly = 0) { resourcePersistencePort.save(any()) }
    }

    @Test
    fun `Given different resource type then update should fail`() {
        sampleMagazine.accessionNumber = sampleAccessionNumber
        sampleBook.accessionNumber = sampleAccessionNumber
        every { resourceSearchPort.findByAccessionNumber(sampleAccessionNumber) } returns sampleMagazine
        Assertions.assertThrows(
            IncompatibleResourceFormatException::class.java
        ) { resourcesService.update(sampleBook) }
        verify(exactly = 0) { resourcePersistencePort.save(any()) }
    }

    @Test
    fun `Given valid resource rd then remove should success`() {
        sampleMagazine.accessionNumber = sampleAccessionNumber
        every { resourceSearchPort.findByAccessionNumber(sampleAccessionNumber) } returns sampleMagazine
        resourcesService.remove(sampleAccessionNumber)
        verify(exactly = 1) { resourcePersistencePort.remove(sampleMagazine) }
    }

    @Test
    fun `Given invalid resource id then remove should fail`() {
        every { resourceSearchPort.findByAccessionNumber(sampleAccessionNumber) } returns null
        Assertions.assertThrows(ResourceNotFoundException::class.java) {
            resourcesService.remove(sampleAccessionNumber)
        }
        verify(exactly = 0) { resourcePersistencePort.remove(any()) }
    }

    @Test
    fun `Given locked resource id then remove should fail`() {
        sampleMagazine.accessionNumber = sampleAccessionNumber
        sampleMagazine.locked = true
        every { resourceSearchPort.findByAccessionNumber(sampleAccessionNumber) } returns sampleMagazine
        Assertions.assertThrows(ResourceBlockedByRentException::class.java) {
            resourcesService.remove(sampleAccessionNumber)
        }
        verify(exactly = 0) { resourcePersistencePort.remove(any()) }
    }

    @Test
    fun `Given valid resource id then getDetails should success`() {
        sampleMagazine.accessionNumber = sampleAccessionNumber
        every { resourceSearchPort.findByAccessionNumber(sampleMagazine.accessionNumber!!) } returns sampleMagazine

        val result = resourcesService.getDetails(sampleMagazine.accessionNumber!!)

        verify(exactly = 1) { resourceSearchPort.findByAccessionNumber(sampleMagazine.accessionNumber!!) }
        Assertions.assertEquals(sampleMagazine, result)
    }

    @Test
    fun `Given invalid resource id then getDetails should fail`() {
        every { resourceSearchPort.findByAccessionNumber(sampleAccessionNumber) } returns null
        Assertions.assertThrows(ResourceNotFoundException::class.java) {
            resourcesService.getDetails(sampleAccessionNumber)
        }
    }
}